recursion/src/recursive_generation.rs

```diff
@@ -0,0 +1,272 @@
+use alloc::vec;
+use alloc::vec::Vec;
+
+use itertools::zip_eq;
+use p3_air::Air;
+use p3_challenger::{CanObserve, CanSampleBits, FieldChallenger, GrindingChallenger};
+use p3_commit::{BatchOpening, Mmcs, Pcs, PolynomialSpace};
+use p3_field::{PrimeCharacteristicRing, TwoAdicField};
+use p3_fri::{FriProof, TwoAdicFriPcs};
+use p3_uni_stark::{
+    Domain, Proof, StarkGenericConfig, SymbolicAirBuilder, Val, VerifierConstraintFolder,
+    get_log_quotient_degree,
+};
+
+#[derive(Debug)]
+pub enum GenerationError {
+    /// Missing parameter for generation,
+    MissingParameterError,
+    /// Too many parameters provided for generation.
+    /// First usize is number of provided parameters,
+    /// second usize is number of expected parameters.
+    TooManyParametersError(usize, usize),
+    /// The FRI batch randomization does not correspond to the ZK setting.
+    RandomizationError,
+}
+
+impl core::fmt::Display for GenerationError {
+    fn fmt(&self, f: &mut core::fmt::Formatter<'_>) -> core::fmt::Result {
+        match self {
+            GenerationError::MissingParameterError => {
+                write!(f, "Missing parameter for challenge generation")
+            }
+            GenerationError::TooManyParametersError(got, expected) => write!(
+                f,
+                "Too many parameters provided for challenge generation: got {}, expected {}",
+                got, expected
+            ),
+            GenerationError::RandomizationError => {
+                write!(
+                    f,
+                    "The FRI batch randomization does not correspond to the ZK setting."
+                )
+            }
+        }
+    }
+}
+
+type ComsWithOpenings<SC> = [(
+    <<SC as StarkGenericConfig>::Pcs as Pcs<
+        <SC as StarkGenericConfig>::Challenge,
+        <SC as StarkGenericConfig>::Challenger,
+    >>::Commitment,
+    Vec<(
+        Domain<SC>,
+        Vec<(
+            <SC as StarkGenericConfig>::Challenge,
+            Vec<<SC as StarkGenericConfig>::Challenge>,
+        )>,
+    )>,
+)];
```

Comment:
Maybe this is cleaner to have:

```suggestion
/// A type alias for a single opening point and its values.
type PointOpening<SC> = (<SC as StarkGenericConfig>::Challenge, Vec<<SC as StarkGenericConfig>::Challenge>);

/// A type alias for all openings within a specific domain.
type DomainOpenings<SC> = Vec<(Domain<SC>, Vec<PointOpening<SC>>)>;

/// A type alias for a commitment and its associated domain openings.
type CommitmentWithOpenings<SC> = (
    <<SC as StarkGenericConfig>::Pcs as Pcs<
        <SC as StarkGenericConfig>::Challenge,
        <SC as StarkGenericConfig>::Challenger,
    >>::Commitment,
    DomainOpenings<SC>,
);

/// The final type alias for a slice of commitments with their openings.
type ComsWithOpenings<SC> = [CommitmentWithOpenings<SC>];
```